src/sql/src/query_model/rewrite_engine.rs

```diff
@@ -967,3 +955,351 @@ impl Rule for WindowingToSelect {
+                    })
+                    .any(|q| {
+                        exactly_one_row(
+                            model,
+                            q.input_box,
+                            exactly_one_row_boxes,
+                            at_most_one_row_boxes,
+                        );
+                        exactly_one_row_boxes[&q.input_box]
+                    })
+            {
+                result = true;
+            }
+            // TODO: 3) a Select box that has a Foreach that meets the
+            //    exactly_one_tuple condition and it is fully joined with the
+            //    remaining Foreach quantifiers via equality predicates with one
+            //    of its unique keys./
+        }
+        BoxType::OuterJoin(_) => {
+            // 4) an OuterJoin box whose PreservedForeach quantifier meets the
+            // exactly_one_tuple condition and its Foreach side (the
+            // non-preserving side) meets at least the one_tuple_at_most
+            // condition.
+            if let Some(preserved_q_id) = b.quantifiers.iter().find(|q_id| {
+                let q = model.get_quantifier(**q_id);
+                exactly_one_row(
+                    model,
+                    q.input_box,
+                    exactly_one_row_boxes,
+                    at_most_one_row_boxes,
+                );
+                q.quantifier_type == QuantifierType::PreservedForeach
+                    && exactly_one_row_boxes[&q.input_box]
+            }) {
+                if b.quantifiers
+                    .iter()
+                    .filter(|q_id| *q_id != preserved_q_id)
+                    .all(|q_id| {
+                        let q = model.get_quantifier(*q_id);
+                        at_most_one_row(
+                            model,
+                            q.input_box,
+                            exactly_one_row_boxes,
+                            at_most_one_row_boxes,
+                        );
+                        at_most_one_row_boxes[&q.input_box]
+                    })
+                {
+                    result = true;
+                }
+            }
+        }
+        BoxType::Values(values) => {
+            // 5) A Values box with one row.
+            if values.rows.len() == 1 {
+                result = true;
+            }
+        }
+        _ => {}
+    }
+    exactly_one_row_boxes.insert(box_id, result);
+}
+
+/// Determines if box corresponding to `box_id` produces at most one row.
+///
+/// Adds the result to `at_most_one_row_boxes`.
+///
+/// Note that this function calls [exactly_one_row].
+///
+/// To save computation effort, the function will refer to information contained in
+/// `at_most_one_row_boxes` and `exactly_one_row_boxes`. Also, if the function
+/// finds out whether or not another box produces at most one row (resp.
+/// exactly one row), it will add that info to `at_most_one_row_boxes` (resp.
+/// `exactly_one_row_boxes`).
+fn at_most_one_row(
```

Comment:
Now that I think about it, this method can be replaced by checking whether the input box has an empty set as its unique key.